### ISSUE TITLE
TimerTrigger Azure Functions and their ParentId

### ISSUE THREAD

**Issue by RomanPavelko** [open-telemetry/opentelemetry-dotnet#4042]
I instrumented opentelemetry for two types of Azure Functions: **HttpTrigger** and **TimerTrigger**. 
The former works fine thanks to `.AddAspNetCoreInstrumentation()` middleware. During its call over HTTP it creates a parent trace which is later shared with other HTTP calls within the function's code and I have a nice nested trace where all function's activities are under their parent.
However with the **TimerTrigger** Azure Function I cannot achieve that probably because `.AddAspNetCoreInstrumentation()` middleware doesn't kick in when TimerTrigger function is executed thus no parent trace is created and all HTTP calls within the function's code behave independently and cannot be grouped under one trace creating a mess.

Here are Grafana screenshots to visualize the problem.

Expected correct result of **HttpTrigger** function:

[![HttpTrigger][1]][1]

Individual non-grouped traces of **TimerTrigger** function:

[![TimerTrigger][2]][2]

Here is the code I have.

**Startup:**

```
using Microsoft.Azure.Functions.Extensions.DependencyInjection;
using Microsoft.Extensions.DependencyInjection;
using OpenTelemetry;
using OpenTelemetry.Logs;
using OpenTelemetry.Metrics;
using OpenTelemetry.Resources;
using OpenTelemetry.Trace;

[assembly: FunctionsStartup(typeof(FunctionsOpenTelemetry.Startup))]
namespace FunctionsOpenTelemetry
{
    public class Startup : FunctionsStartup
    {
        public override void Configure(IFunctionsHostBuilder builder)
        {
            // OpenTelemetry Resource to be associated with traces
            var openTelemetryResourceBuilder = ResourceBuilder.CreateDefault().AddService("Reporting-Serivce");

            // Enable Tracing with OpenTelemetry
            var openTelemetryTracerProvider = Sdk.CreateTracerProviderBuilder()
                .SetResourceBuilder(openTelemetryResourceBuilder)
                .SetSampler(new AlwaysOnSampler())
                .AddHttpClientInstrumentation(g => g.Filter = context => !context.RequestUri.ToString().Contains("127.0.0.1"))
                .AddAspNetCoreInstrumentation()
                .AddConsoleExporter()
                .AddJaegerExporter()
                .Build();
            builder.Services.AddSingleton(openTelemetryTracerProvider);
        }
    }
}
```

**Functions code:**

```
using System;
using System.Threading.Tasks;
using Microsoft.Azure.WebJobs;
using Microsoft.Azure.WebJobs.Extensions.Http;
using Microsoft.AspNetCore.Http;
using Microsoft.Extensions.Logging;
using System.Net.Http;

namespace FunctionsOpenTelemetry
{
    public static class Function1
    {
        [FunctionName("Function1")]
        public static async Task Run1([HttpTrigger(AuthorizationLevel.Anonymous, "get", "post", Route = null)] HttpRequest req, ILogger log)
        {
            DoTenHttpRequests();
        }

        [FunctionName("Function2")]
        public static async Task Run2([TimerTrigger("0 */5 * * * *")] TimerInfo myTimer, ILogger log)
        {
            DoTenHttpRequests();
        }

        private static void DoTenHttpRequests()
        {
            for (var i = 0; i < 10; i++)
            {
                var client = new HttpClient();
                client.BaseAddress = new Uri("https://www.google.com");
                client.GetAsync("/").GetAwaiter().GetResult();
            }
        }
    }
}
```

**.csproj:**

```
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <TargetFramework>net6.0</TargetFramework>
    <AzureFunctionsVersion>v4</AzureFunctionsVersion>
  </PropertyGroup>
  <ItemGroup>
    <PackageReference Include="Microsoft.Azure.Functions.Extensions" Version="1.1.0" />
    <PackageReference Include="Microsoft.NET.Sdk.Functions" Version="4.0.1" />
    <PackageReference Include="OpenTelemetry.Exporter.Console" Version="1.2.0-rc2" />
    <PackageReference Include="OpenTelemetry.Instrumentation.AspNetCore" Version="1.0.0-rc9" />
	<PackageReference Include="OpenTelemetry.Instrumentation.Http" Version="1.0.0-rc9" />
	<PackageReference Include="OpenTelemetry.Exporter.Jaeger" Version="1.2.0-rc2" />
  </ItemGroup>
  <ItemGroup>
    <None Update="host.json">
      <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
    </None>
    <None Update="local.settings.json">
      <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
      <CopyToPublishDirectory>Never</CopyToPublishDirectory>
    </None>
  </ItemGroup>
</Project>

```

Summarizing my question: how can I instrument **TimerTrigger** Azure Functions so they create an initial parent trace and share it within their code like **HttpTrigger** function does?

  [1]: https://i.stack.imgur.com/DnqEs.png
  [2]: https://i.stack.imgur.com/g71n9.png

**Comment by RomanPavelko** [comment]
I found a workaround how to make a **TimerTrigger** function create a parent span and have all other child activities logged under that parent span. Here is the code (**see Run2 method**):


    public class Function1
    {
        private readonly TracerProvider openTelemetryTracerProvider;
        
        public Function1(TracerProvider openTelemetryTracerProvider)
        {
            this.openTelemetryTracerProvider = openTelemetryTracerProvider;
        }

        [FunctionName("Function1")]
        public async Task Run1([HttpTrigger(AuthorizationLevel.Anonymous, "get", "post", Route = null)] HttpRequest req, ILogger log)
        {
            DoTenHttpRequests();
        }

        [FunctionName("Function2")]
        public async Task Run2([TimerTrigger("0 */1 * * * *")] TimerInfo myTimer, ILogger log)
        {
            var tracer = this.openTelemetryTracerProvider.GetTracer("OpenTelemetry.Instrumentation.AspNetCore");
            using var scope = tracer.StartActiveSpan("my-custom-span");

            DoTenHttpRequests();
        }

        private static void DoTenHttpRequests()
        {
            for (var i = 0; i < 10; i++)
            {
                var client = new HttpClient();
                client.BaseAddress = new Uri("https://www.google.com");
                client.GetAsync("/").GetAwaiter().GetResult();
            }
        }
    }

Basically I'm retrieving registered in `Startup.cs` class `OpenTelemetry.Instrumentation.AspNetCore` tracer, manually starting this span and have all function's logic under `using` statement of this span.

I briefly checked `OpenTelemetry.Instrumentation.AspNetCore` implementation and it looks like it only listens to incoming HTTP requests where **TimerTrigger** assuming is not such a request thus no parent span is created. Is there a plan to extend `OpenTelemetry.Instrumentation.AspNetCore` instrumentation for supporting **TimerTrigger** functions or, perhaps, there is a plan to create a completely new instrumentation specifically for **TimerTrigger** functions?

**Comment by alanwest** [comment]
> Is there a plan to extend OpenTelemetry.Instrumentation.AspNetCore instrumentation for supporting TimerTrigger functions

No, this will likely not be the plan. The AspNetCore instrumentation is not implemented with Azure Functions in mind. It's a coincidence that Azure Functions uses some of the foundation of ASP.NET Core thereby invoking the instrumentation. This is true only for some use cases - namely HttpTrigger as you've discovered, and specifically when using the in-process hosting model for Azure Functions not the isolated worker process. See https://learn.microsoft.com/en-us/azure/azure-functions/dotnet-isolated-in-process-differences.

> or, perhaps, there is a plan to create a completely new instrumentation specifically for TimerTrigger functions?

Azure Functions specific instrumentation is more likely. Though, to my knowledge, no one is currently working on this. A contribution of Azure Function instrumentation would be most appropriate for the [opentelemetry-dotnet-contrib project](https://github.com/open-telemetry/opentelemetry-dotnet-contrib).

> I found a workaround

👍 Yep manual instrumentation is a good solution. Just some a recommendations:

* Use `ActivitySource` directly rather that using `TracerProvider`.
* Do not use the name `OpenTelemetry.Instrumentation.AspNetCore` for your ActivitySource.

So something like:

```csharp
public class Function1
{
    private static ActivitySource activitySource = new ActivitySource("MyActivitySource");

    [FunctionName("Function2")]
    public async Task Run2([TimerTrigger("0 */1 * * * *")] TimerInfo myTimer, ILogger log)
    {
        using var activity = activitySource.StartActivity("my-custom-span");
        activity?.SetTag("some-tag", "some-value");
        DoTenHttpRequests();
    }
}
```

This would require you do subscribe to MyActivitySource when configuring your tracer provider

```csharp
var openTelemetryTracerProvider = Sdk.CreateTracerProviderBuilder()
                .SetResourceBuilder(openTelemetryResourceBuilder)
                .SetSampler(new AlwaysOnSampler())
                .AddHttpClientInstrumentation(g => g.Filter = context => !context.RequestUri.ToString().Contains("127.0.0.1"))
                .AddSource("MyActivitySource")
                .AddConsoleExporter()
                .AddJaegerExporter()
                .Build();
```

**Comment by RomanPavelko** [comment]
@alanwest understood. Many thanks for answering my questions and for the hints on how to proceed with TimerTrigger functions. I will be looking forward to any updates for the out of the box solution for these types of functions and as for now I'll go with manual instrumentation.
